Docs for multi-tenancy. (#620)
timescale/promscale: docs/multi_tenancy.md

Promscale keeps tenant name as `__tenant__: <tenant_name>` label pair attached to the series that is
ingested in that tenant's write-request.

### Using headers

Tenant name can be supplied with a header with the key `TENANT` and the value corresponding the tenant name.

Example:

`TENANT: tenant-A`

In Prometheus remote-write config

```yaml
remote_write:
  - url: http://localhost:9201/write
    headers:
      TENANT: A
```

### Using external_labels

Promscale reserves `__tenant__` as label_key for storing tenant name for the incoming series. You can
push data directly with this label as

```yaml
global:
  external_labels:
    __tenant__: A
```

## Querying multi-tenant data

Since data from any tenant is kept with a `__tenant__` label key, it can be used to query any tenant data.
Please note only those tenants are allowed in the query, that is authorized under `-multi-tenancy-valid-tenants`
flag. Moreover, non-tenants are allowed to be queried only if `-multi-tenancy-allow-non-tenants` is applied.

If a query aims to be evaluated across multiple tenants (say tenant-A and tenant-B), it can be done
by `metric_name{__tenant__=~"tenant-A|tenant-B"}` respectively.

Example:

> Imagine if Promscale contains data from `tenant-A`, `tenant-B` and `tenant-C`. If Promscale is applied with

Reviewer: ```suggestion
Imagine if Promscale contains data from `tenant-A`, `tenant-B` and `tenant-C`. If Promscale is configured with
```

`-multi-tenancy-valid-tenants=tenant-A,tenant-B` then,

If the PromQL query is
1. `metric_name{__tenant__=tenant-A}` then data corresponding to `metric_name` from `tenant-A` will
    only be returned.
2. `metric_name{__tenant__=tenant-C}` then no data will be returned.
3. `metric_name{__tenant__=~tenant-A|tenant-B}` then data from `metric_name` from tenants `tenant-A` and `tenant-B`
    will be returned.
4. `metric_name{__tenant__=~tenant-A|tenant-C}` then data from `metric_name` will be only from `tenant-A` and

>     not from `tenant-C`. This is because `tenant-C` is not a valid tenant for the current session of Promscale.

Reviewer: ```suggestion
    not from `tenant-C`. This is because the Promscale instance is configured to return data from only `tenant-A` and `tenant-B`.
```

5. `metric_name` then the data from `metric_name` will be from `tenant-A` and `tenant-B` only.

> Now, if the same database also contains data from non-tenants and `-multi-tenancy-allow-non-tenants` is applied then

Reviewer: ```suggestion
Now, if the same database also contains samples without any tenant label and `-multi-tenancy-allow-non-tenants` is applied then
```

If the PromQL query is
1. `metric_name{__tenant__=tenant-A}` then data corresponding to `metric_name` from `tenant-A` will
   only be returned.
2. `metric_name{__tenant__=tenant-A|}` then data corresponding to `metric_name` from non-tenant and from `tenant-A`
   will be returned.

> 3. `metric_name` then data from `metric_name` will be from the non-tenants, `tenant-A` and `tenant-B` respectively.

Reviewer 2: Add another example to query non-tenant data only?

querying its own data while being sure that it could not access other tenant's data.

Multi-tenancy in Promscale features:
1. Simple configuration for multi-tenancy using headers or external labels
2. Cross-tenant queries in PromQL and SQL
3. Allow data with no tenant information to be written or queried when multi-tenant mode is enabled
4. Restricting the set of valid tenants a Promscale instance can ingest or query
5. Authorization of tenants using bearer_tokens

## Configuring Promscale for multi-tenancy

By default, Promscale ingests data without multi-tenancy. However, configuring Promscale for multi-tenancy is simple.
For Promscale to ingest tenant data, apply the `-multi-tenancy` flag during the start. This will allow Promscale to
ingest data from all tenants by default. If you want Promscale to ingest or query data from a restricted list of tenants, mention the
tenant names in `-multi-tenancy-valid-tenants=<tenant_name_seperated_by_commas>`.

When multi-tenancy is enabled, accepting data without a tenant name being specified is disabled by default.
If you want to accept data without the tenant information, apply the `-multi-tenancy-allow-non-tenants` flag.

> Remember: configuration flags are for a Promscale instance only. If you have more than one Promscale instance,
> you will require applying flags on other Promscale instances as well (If you want them to participate in multi-tenant operations).

Reviewer: ```suggestion
Remember to set the appropriate CLI flags on all of your promscale instances.
```
